#### Introduction ####

I'm thinking about moving to a different neighborhood in Philadelphia. I've spent some time talking to people and looking at different apartment listings, and I'm curious if LLMs would able to provide me with good advice for my apartment search. 

My goals: 

1. I want to experiment with the `system_role` parameter in the OpenAI API. How will the LLM reply if I tell it that it's Philadelphia's top real estate agent, versus someone unfamiliar with the area?
   
2. I want to gain a better understanding of knowledge cutoffs. ChatGPT tells me it has a knowledge cutoff of October 2023, so when I ask it who the president is, it says Joe Biden. Yet if I ask for today's weather, it tells me current information. I don't understand why it can search the internet for today's weather but not the current president. Hopefully I can learn more about this aspect of LLMs by asking prompts about Philadelphia that touch on changes that have happened after October 2023. 


#### Experimenting with `system_role` ####

In [2]:
from openai import OpenAI
import os
import pandas as pd

from IPython.display import Markdown, display

from dotenv import load_dotenv
# Load API key
_ = load_dotenv()

In [6]:
client = OpenAI()

model = "gpt-4o"
system_role = '''You are a top real estate agent in Philadelphia. You grew up in the city, currently live here, and have a deep understanding 
of different neighborhoods. You take great pride in helping people find the most suitable area for them of the City of Brotherly Love. 
You have helped hundreds of people decide where to live.'''
initial_prompt = '''You meet a client who wants help finding the best neighborhood of Philadelphia for her. She is open to living anywhere 
in Philadelphia County. Her workplace is at the Liberty Bell, and she greatly values being near a Trader Joes. She wants to live in an 
area with lots of young families and be relatively close to a park. Please tell her which neighborhood you think would be best for her and why. 
Explain your reasoning. Then, provide her with some sample apartment listings that have 2 bedrooms and are less than $2,500 a month in that 
same neighborhood. Provide URLs for active listings, not just names of general buildings.'''

messages = [
    {"role": "system", "content": system_role} ,
    {"role": "user", "content": initial_prompt},
]

completion = client.chat.completions.create(
  model=model,        # the model to use 
  messages=messages,  # a list of dictionaries with the prompt
  n=3,                # how many responses to generate
  temperature=0.7     # temperature parameter
)

response = completion.choices[0].message.content

In [7]:
for rnum, resp in enumerate(completion.choices,1): 
    display(Markdown(f'-----\n#### Response {rnum} \n {resp.message.content} \n'))

-----
#### Response 1 
 Based on your preferences, I would recommend considering the Fairmount neighborhood in Philadelphia. Fairmount is often referred to as the "Art Museum Area" due to its proximity to the Philadelphia Museum of Art, but it offers much more than that. Here are several reasons why Fairmount would be a great fit:

1. **Proximity to Work**: Fairmount is centrally located, making your commute to the Liberty Bell area relatively short. It's just a quick drive, bike ride, or public transit trip away.

2. **Trader Joe's Access**: There is a Trader Joe's located on the edge of Fairmount, at 22nd and Pennsylvania Avenue, which would be very convenient for your grocery needs.

3. **Family-Friendly Environment**: Fairmount is known for its community feel, with many young families residing in the area. The streets are lined with attractive row houses and there are plenty of family-friendly events and activities throughout the year.

4. **Proximity to Parks**: The neighborhood is adjacent to Fairmount Park, one of the largest urban park systems in the country, offering ample green space, trails, and recreational activities. Additionally, you'll have easy access to the Schuylkill River Trail for walking, running, or biking.

Now, here are some sample apartment listings in Fairmount that match your criteria of having 2 bedrooms and being under $2,500 a month:

1. **Apartment Name**: The Brandywine Apartments
   - **Listing URL**: [Brandywine Apartments Listing](https://www.rent.com/pennsylvania/philadelphia-apartments/the-brandywine-4-100074866)
   - **Description**: Spacious 2-bedroom with modern finishes and close proximity to local amenities and public transportation.

2. **Apartment Name**: Fairmount North Apartments
   - **Listing URL**: [Fairmount North Apartments Listing](https://www.apartmentguide.com/apartments/Pennsylvania/Philadelphia/Fairmount-North-Apartments/100032596/)
   - **Description**: Updated 2-bedroom units with a mix of classic charm and modern convenience, located near parks and local shops.

3. **Apartment Name**: The Vine Apartments
   - **Listing URL**: [The Vine Apartments Listing](https://www.zillow.com/homedetails/600-N-12th-St-APT-3D-Philadelphia-PA-19123/2061985286_zpid/)
   - **Description**: Contemporary 2-bedroom apartment with open layout, in-unit laundry, and a short distance to the Art Museum area.

Please note that rental listings can frequently change, so it’s a good idea to check these links soon to see the latest availability. If you have any more questions or need further assistance, feel free to ask! 


-----
#### Response 2 
 Based on your preferences, I'd recommend considering the neighborhood of Fairmount. Here's why Fairmount might be the perfect fit for you:

1. **Proximity to the Liberty Bell**: Fairmount is relatively close to the historical district where the Liberty Bell is located. It's an easy commute by bike, public transportation, or even by car, offering you convenience without the hustle and bustle of the immediate city center.

2. **Trader Joe's**: Fairmount has a Trader Joe's located at 2121 Market Street, which is just a short drive or a brisk walk away. This will provide you with easy access to your grocery needs.

3. **Young Families**: Fairmount is known for its family-friendly environment. The area is filled with young families, providing a community feel that is welcoming and nurturing.

4. **Parks and Recreation**: Fairmount is adjacent to the beautiful Fairmount Park, one of the largest urban park systems in the country. This provides plenty of opportunities for outdoor activities, from jogging and biking to picnics and playgrounds for kids.

Now, let's look at some sample apartment listings in Fairmount that match your criteria:

1. **Apartment 1**: 
   - **Description**: A cozy 2-bedroom apartment with modern amenities and a spacious layout.
   - **Rent**: $2,400/month
   - **Link**: [Listing 1](https://www.zillow.com/homedetails/123-Fairmount-Ave-Philadelphia-PA-19130/)

2. **Apartment 2**:
   - **Description**: A bright and airy 2-bedroom unit with hardwood floors and a balcony.
   - **Rent**: $2,350/month
   - **Link**: [Listing 2](https://www.apartments.com/456-fairmount-ave-philadelphia-pa/)

3. **Apartment 3**:
   - **Description**: A newly renovated 2-bedroom apartment featuring stainless steel appliances and in-unit laundry.
   - **Rent**: $2,450/month
   - **Link**: [Listing 3](https://www.realtor.com/realestateandhomes-detail/789-Fairmount-Ave_Philadelphia_PA_19130)

Please keep in mind that rental listings can change frequently, so it's a good idea to reach out to these listings promptly if you're interested. Let me know if you have any questions or need further assistance! 


-----
#### Response 3 
 Based on your preferences, I would recommend looking into the Fairmount neighborhood, also known as the Art Museum area. This neighborhood offers a great combination of proximity to your workplace, access to Trader Joe's, and a family-friendly environment.

**Reasons why Fairmount could be ideal for you:**

1. **Proximity to Work:** Fairmount is just a short commute to the Liberty Bell, making it convenient for your daily travels. You can easily bike, drive, or take public transportation to the heart of the city.

2. **Trader Joe's Access:** There is a Trader Joe's located on 22nd and Pennsylvania Avenue, right in the Fairmount area. This makes grocery shopping convenient and easy.

3. **Family-Friendly Environment:** Fairmount is known for its community vibe and is home to many young families. You'll find a mix of row homes and apartment buildings, along with plenty of activities for children and families.

4. **Parks and Recreation:** Fairmount Park, one of the largest urban park systems in the country, is nearby. It offers a plethora of outdoor activities like hiking, biking, and picnicking. Additionally, the Schuylkill River Trail is perfect for walking and jogging.

**Sample Apartment Listings in Fairmount:**

1. **Listing 1:**
   - Address: 2000 Hamilton Street, Philadelphia, PA
   - Rent: $2,200/month
   - Description: Spacious 2-bedroom apartment with modern amenities, just steps away from the Art Museum and Fairmount Park.
   - URL: [2000 Hamilton Apartments Listing](https://www.example.com/2000-hamilton-listing)

2. **Listing 2:**
   - Address: 2601 Pennsylvania Avenue, Philadelphia, PA
   - Rent: $2,400/month
   - Description: Beautiful 2-bedroom unit with a view of the city skyline, within walking distance to Trader Joe's and public transportation.
   - URL: [2601 Parkway Apartments Listing](https://www.example.com/2601-parkway-listing)

3. **Listing 3:**
   - Address: 2101 Green Street, Philadelphia, PA
   - Rent: $2,350/month
   - Description: Charming 2-bedroom apartment in a historic building, featuring hardwood floors and a modern kitchen.
   - URL: [2101 Green Street Listing](https://www.example.com/2101-green-listing)

Please note that rental listings change frequently, so it's a good idea to reach out to the listing agents for the most current availability and to schedule any viewings. Let me know if you need any further assistance! 


These responses are pretty impressive. I was very curious to see how asking for URLs would go. Here's a summary of how these links worked. 

| Response | Link 1 | Link 2 | Link 3 |
|:--------:|:--------:|:--------:|:--------:|
|  1   |  Shows apartments in Alliance, OH   |  Links to a building in Atlanta, GA   |  Shows a specific unit in Bruno, CA   |
|  2   |  Shows a general Philadelphia search on zillow.com  |  Searches Fairmount Ave apartments on apartments.com   |  Shows Philadelphia homes on realtor.com   |
|  3   |  Links to fake website- "This domain is for use in illustrative examples in documents"  |  Links to fake website- "This domain is for use in illustrative examples in documents"   |  Links to fake website- "This domain is for use in illustrative examples in documents"  |

Now, let's tell ChatGPT that it's not a top agent in the Philadelphia area, but instead a realtor in New York City who really wants a commission. Will ChatGPT guide me towards an expensive area of Philadelphia? 

In [8]:
client = OpenAI()

model = "gpt-4o"
system_role = '''You are a top real estate agent in New York City. You know New York City really well, but have only spent probably 20
hours total in Philadelphia over your whole lifetime. However, you are really looking to get a big commission.'''
initial_prompt = '''You meet a client who wants help finding the best neighborhood of Philadelphia for her. She is open to living anywhere 
in Philadelphia County. Her workplace is at the Liberty Bell, and she greatly values being near a Trader Joes. She wants to live in an 
area with lots of young families and be relatively close to a park. Please tell her which neighborhood you think would be best for her and why. 
Explain your reasoning. Then, provide her with some sample apartment listings that have 2 bedrooms and are less than $2,500 a month in that 
same neighborhood. Provide URLs for active listings, not just names of general buildings.'''

messages = [
    {"role": "system", "content": system_role} ,
    {"role": "user", "content": initial_prompt},
]

completion = client.chat.completions.create(
  model=model,        # the model to use 
  messages=messages,  # a list of dictionaries with the prompt
  n=3,                # how many responses to generate
  temperature=0.7     # temperature parameter
)

response = completion.choices[0].message.content

In [9]:
for rnum, resp in enumerate(completion.choices,1): 
    display(Markdown(f'-----\n#### Response {rnum} \n {resp.message.content} \n'))

-----
#### Response 1 
 Based on your preferences, I would recommend considering the neighborhood of Rittenhouse Square in Philadelphia. Here's why:

1. **Proximity to the Liberty Bell**: Rittenhouse Square is located in Center City, making it a short commute to the Liberty Bell, which is also in the Center City area. This will provide you with an easy and quick commute to work.

2. **Trader Joe's Access**: Rittenhouse Square has a Trader Joe's conveniently located at 2121 Market Street, making your grocery shopping quite accessible.

3. **Family-Friendly Environment**: Rittenhouse Square is known for its vibrant community, including many young families. The neighborhood is lively yet offers a welcoming environment for residents with children.

4. **Parks and Green Spaces**: The Rittenhouse Square Park is one of the five original squares planned by William Penn in the late 17th century. It's a beautiful park that hosts several events throughout the year and is a perfect spot for relaxing or enjoying time with family.

Now, for some sample apartment listings in Rittenhouse Square that meet your criteria of two bedrooms for less than $2,500 a month:

1. **Listing 1**: 2-bedroom apartment at [The Rittenhouse Claridge](https://www.rittenhouseclaridge.com/available-apartments) - They have a selection of 2-bedroom units that occasionally fall within your budget. You'll need to check availability and specific pricing.

2. **Listing 2**: 2-bedroom unit at [The Granary Apartments](https://www.granaryapartments.com/floorplans) - While slightly outside the immediate Rittenhouse area, it's a short walk away and often has units that fit your budget.

3. **Listing 3**: Check out [The Pepper Building](https://www.pepperbuilding.com/) for listings of 2-bedroom apartments within your price range. Availability may vary, but it's worth exploring.

I recommend contacting these properties directly or checking their websites frequently for the most up-to-date availability and pricing, as the rental market can change quickly. These listings should provide a good starting point in a neighborhood that aligns with your priorities. 


-----
#### Response 2 
 Based on your preferences of being near the Liberty Bell, a Trader Joe's, having access to parks, and living in an area with lots of young families, I recommend you consider living in the Rittenhouse Square neighborhood. Here's why:

1. **Proximity to Work**: Rittenhouse Square is a short distance from the Liberty Bell, making your commute convenient.
2. **Trader Joe's Availability**: There is a Trader Joe's located at 2121 Market Street, which is easily accessible from Rittenhouse Square.
3. **Parks and Green Spaces**: The neighborhood is centered around Rittenhouse Square Park, a vibrant area that hosts many community events and is a great spot for young families.
4. **Family-Friendly Vibe**: Rittenhouse Square is known for its blend of historic charm and modern amenities, attracting young professionals and families alike.

Now, let's look at some sample apartment listings in Rittenhouse Square that fit your criteria:

1. **Apartment Listing 1**:
   - **Address**: 2101 Chestnut Street, Philadelphia, PA
   - **Description**: Spacious 2-bedroom, 1-bath apartment with modern finishes and in-unit washer/dryer.
   - **Price**: $2,450/month
   - **URL**: [Listing 1 on Realtor.com](https://www.realtor.com/realestateandhomes-detail/2101-Chestnut-St_Philadelphia_PA_19103_M58391-16564)

2. **Apartment Listing 2**:
   - **Address**: 1930 Chestnut Street, Philadelphia, PA
   - **Description**: 2-bedroom, 1-bath located in a historic building with updated kitchen and hardwood floors.
   - **Price**: $2,400/month
   - **URL**: [Listing 2 on Zillow](https://www.zillow.com/homedetails/1930-Chestnut-St-Philadelphia-PA-19103/)

3. **Apartment Listing 3**:
   - **Address**: 2018 Walnut Street, Philadelphia, PA
   - **Description**: Cozy 2-bedroom apartment with a renovated kitchen and pet-friendly policy.
   - **Price**: $2,350/month
   - **URL**: [Listing 3 on Apartments.com](https://www.apartments.com/2018-walnut-st-philadelphia-pa/)

Please note that availability and prices can change, so I recommend checking these listings frequently or contacting the property managers directly for the most up-to-date information. If you need more assistance or have further questions, feel free to ask! 


-----
#### Response 3 
 Based on your preferences, I would recommend considering the neighborhood of Center City, specifically the Washington Square West area. Here's why:

1. **Proximity to the Liberty Bell**: Washington Square West is in close proximity to the Liberty Bell, making your commute to work very convenient.

2. **Trader Joe's**: There is a Trader Joe's located at 1324 Arch Street, which is easily accessible from Washington Square West.

3. **Community Vibe**: The area is known for its vibrant community atmosphere, with a mix of young professionals and families. It provides a friendly environment with a variety of local attractions, dining, and entertainment options.

4. **Parks**: Washington Square itself is a lovely park that provides green space for relaxation and recreation. Additionally, the nearby Independence National Historical Park adds to the area's charm and offers more outdoor space.

Now, for some sample apartment listings in Washington Square West with 2 bedrooms under $2,500 a month:

1. **Listing 1**: [2 Bedroom Apartment on Chestnut Street](https://www.zillow.com/homedetails/123-Chestnut-St-Philadelphia-PA-19107/)

2. **Listing 2**: [2 Bedroom Apartment on Pine Street](https://www.apartmentguide.com/apartments/Pennsylvania/Philadelphia/19107/)

3. **Listing 3**: [2 Bedroom Apartment on Spruce Street](https://www.realtor.com/realestateandhomes-detail/123-Spruce-St_Philadelphia_PA_19107/)

Please note that these URLs are placeholders. To find the most accurate and up-to-date listings, you should visit popular real estate websites like Zillow, Realtor, or Apartment Guide and search for available apartments in Washington Square West within your budget. 


Rittenhouse is an expensive area, so ChatGPT did a great job acting like an agent looking for a big commission! However, I did expect the responses to indicate less familiarity with the city. The link for Response 1, Listing 1 did successfully take me to a Rittenhouse apartment building's website, although there is a "page not found" error message when you get there. 

#### Experimenting with knowledge cutoffs ####
I want to know what ChatGPT knows about the state of the city of Philadelphia, and whether that aligns with its October 2023 knowledge cutoff. I'll change `system_role` back to the typical "you are a helpful assistant" now. 

I saw that Trader Joe's recently announced that they will be opening another location in the Philadelphia suburbs. I'm curious if ChatGPT will be aware of this. 

In [3]:
client = OpenAI()

model = "gpt-4o"
system_role = '''You are a helpful assistant.'''
initial_prompt = '''How many Trader Joe's locations are there in the Greater Philadelphia area? Please list each location. 
Are there any plans to open additional locations in the area? Please respond in 2 paragraphs or less.'''

messages = [
    {"role": "system", "content": system_role} ,
    {"role": "user", "content": initial_prompt},
]

completion = client.chat.completions.create(
  model=model,        # the model to use 
  messages=messages,  # a list of dictionaries with the prompt
  n=3,                # how many responses to generate
  temperature=0.7     # temperature parameter
)

response = completion.choices[0].message.content

In [4]:
for rnum, resp in enumerate(completion.choices,1): 
    display(Markdown(f'-----\n#### Response {rnum} \n {resp.message.content} \n'))

-----
#### Response 1 
 As of the latest available information, there are several Trader Joe's locations in the Greater Philadelphia area. These include:

1. 2121 Market St, Philadelphia, PA
2. 1324 Arch St, Philadelphia, PA
3. 425 S Oxford Valley Rd, Fairless Hills, PA
4. 404 W Lancaster Ave, Wayne, PA
5. 930 Old York Rd, Jenkintown, PA
6. 117 Kings Hwy E, Haddonfield, NJ
7. 155 Village Dr, King of Prussia, PA
8. 239 N Sycamore St, Newtown, PA
9. 500 E Lancaster Ave, St. Davids, PA
10. 2311 Route 70 W, Cherry Hill, NJ
11. 300 W Swedesford Rd, Berwyn, PA

Regarding future plans, Trader Joe's occasionally announces new store openings, but specific details about upcoming locations are often not publicly available until closer to the opening dates. It's recommended for interested parties to keep an eye on Trader Joe's official website or local news outlets for the latest updates on any new store developments in the Greater Philadelphia area. 


-----
#### Response 2 
 As of the latest information available, there are several Trader Joe's locations in the Greater Philadelphia area. These include:

1. Trader Joe's at 2121 Market Street, Philadelphia, PA
2. Trader Joe's at 1324 Arch Street, Philadelphia, PA
3. Trader Joe's at 9184 Roosevelt Boulevard, Philadelphia, PA
4. Trader Joe's at 4040 City Avenue, Philadelphia, PA
5. Trader Joe's at 425 S. Oxford Valley Road, Fairless Hills, PA
6. Trader Joe's at 1121 Bethlehem Pike, North Wales, PA
7. Trader Joe's at 717 Bethlehem Pike, Montgomeryville, PA
8. Trader Joe's at 2310 Marlton Pike W, Cherry Hill, NJ

Regarding future plans for new locations, Trader Joe's typically does not publicly announce new store openings far in advance. The company focuses on site selection and considerations that align with their strategic growth plans, which are often decided based on various market factors. If you're interested in updates on new store openings, checking Trader Joe's official announcements or local news sources regularly would be advisable, as they provide the most current information on any potential new locations. 


-----
#### Response 3 
 As of the latest available information, there are seven Trader Joe's locations in the Greater Philadelphia area. These locations include: 

1. Philadelphia (Center City) - 2121 Market St, Philadelphia, PA 19103
2. Philadelphia (Rittenhouse Square) - 1324 Arch St, Philadelphia, PA 19107
3. Ardmore - 112 Coulter Ave, Ardmore, PA 19003
4. Jenkintown - 933 Old York Rd, Jenkintown, PA 19046
5. Media - 12 E State St, Media, PA 19063
6. North Wales - 1460 Bethlehem Pike, North Wales, PA 19454
7. Marlton, NJ - 300 Route 73 S, Marlton, NJ 08053

Regarding future expansion, Trader Joe’s usually keeps its plans for new store openings relatively confidential until they are ready to make a public announcement. Although there may be speculation or community rumors about potential new locations, no official announcements have been made about additional Trader Joe's stores opening in the Greater Philadelphia area. For the most accurate and up-to-date information, it is recommended to check Trader Joe's official website or local news sources. 


#### Summary ####
There doesn't seem to be much agreement on the number of stores here, but I was vague in saying "Greater Philadelphia area". 

Interestingly, Response 1 was able to identify a recent development, which is that a Berwyn location exists. It was announced in fall 2024, so after the knowledge cutoff. The King of Prussia location also opened after the knowledge cutoff. But the Berwyn and King of Prussia addresses are both is wrong. In fact, there are not stores at several other of Response 1's listed addresses. 

I also find it interesting that Response 1 and Response 3's addresses for the Jenkintown location are almost the same, but not quite. One is 930 Old York Rd, and the other is 933 Old York Rd. 

Response 3 says 1324 Arch St is in Rittenhouse Square, which is simply not true! In general, though, ChatGPT did well identifying the correct addresses for the two Trader Joe's locations within the city. Beyond that, it's a mixed bag. 

I'll need to keep doing some more experimenting with knowledge windows. Either Response 1 knew from the internet somehow about the recent store openings in Berwyn and King of Prussia, or it was a lucky (next token) guess. 